controller/instance_manager_controller.go

```diff
@@ -2044,6 +2087,265 @@ func (m *InstanceManagerMonitor) StopMonitorWithLock() {
+			m.logger.WithError(err).Errorf("Failed to check orphan for replica instance %v", replicaInstance.Spec.Name)
+		} else if isOrphaned {
+			m.logger.Infof("Replica instance %v is orphaned", replicaInstance.Spec.Name)
+			err := m.createOrphan(im, replicaInstance.Spec.Name, longhorn.OrphanTypeReplicaInstance, replicaInstance.Spec.DataEngine)
+			if err != nil {
+				m.logger.WithError(err).Errorf("Failed to create orphan for replica instance %v", replicaInstance.Spec.Name)
+			}
+		}
+	}
+
+	// update instance state on orphan CRs
+	deleteOrphans := make(map[string]bool)
+	for orphanName, orphan := range existOrphans {
+		instanceName := orphan.Spec.Parameters[longhorn.OrphanInstanceName]
+		var instanceProc longhorn.InstanceProcess
+		var instanceExist, instanceCRScheduledBack bool
+		switch orphan.Spec.Type {
+		case longhorn.OrphanTypeEngineInstance:
+			instanceProc, instanceExist = engineProcess[instanceName]
+			isCRScheduledBack, err := m.isEngineScheduledOnNode(instanceName)
+			if err != nil {
+				m.logger.WithError(err).WithField("instance", instanceName).Errorf("Failed to check CR scheduled node for engine instance orphan %v", orphanName)
+				instanceCRScheduledBack = false
+			} else {
+				instanceCRScheduledBack = isCRScheduledBack
+			}
+		case longhorn.OrphanTypeReplicaInstance:
+			instanceProc, instanceExist = replicaProcess[instanceName]
+			isCRScheduledBack, err := m.isReplicaScheduledOnNode(instanceName)
+			if err != nil {
+				m.logger.WithError(err).WithField("instance", instanceName).Errorf("Failed to check CR scheduled node for replica instance orphan %v", orphanName)
+				instanceCRScheduledBack = false
+			} else {
+				instanceCRScheduledBack = isCRScheduledBack
+			}
+		}
+		var instanceState longhorn.InstanceState
+		if instanceExist {
+			instanceState = instanceProc.Status.State
+		} else {
+			instanceState = longhorn.InstanceStateTerminated
+		}
+
+		if orphan.DeletionTimestamp.IsZero() {
+			if !instanceExist || instanceCRScheduledBack {
+				m.logger.WithFields(logrus.Fields{
+					"instanceExist":           instanceExist,
+					"instanceState":           instanceState,
+					"instanceCRScheduledBack": instanceCRScheduledBack,
+				}).Debugf("Orphan instance %s is deletable", orphanName)
+				deleteOrphans[orphanName] = true
+			}
+		}
+	}
+
+	// delete orphan CRs
+	autoDeletionTypes, err := m.ds.GetSettingOrphanResourceAutoDeletion()
+	var autoDeleteEngine = false
+	var autoDeleteReplica = false
+	if err != nil {
+		m.logger.WithError(err).Warnf("Failed to fetch orphan auto deletion setting, disable by default")
+	} else {
+		autoDeleteEngine = autoDeletionTypes[types.OrphanResourceTypeEngineInstance]
+		autoDeleteReplica = autoDeletionTypes[types.OrphanResourceTypeReplicaInstance]
+	}
+	m.cleanupOrphans(existOrphans, deleteOrphans, autoDeleteEngine, autoDeleteReplica)
+}
+
+func (m *InstanceManagerMonitor) getNodeRuntimeOrphanMap() (map[string]*longhorn.Orphan, error) {
```

Comment:
It is not easy to quickly understand the purpose of the function from the function name.
Suggest `listOrphansForEngineAndReplicaInstances(nodeName string)` or similar one.